src/org/jetbrains/java/decompiler/struct/JarContextSource.java

```diff
@@ -0,0 +1,135 @@
+// Copyright 2000-2022 JetBrains s.r.o. and ForgeFlower contributors Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package org.jetbrains.java.decompiler.struct;
+
+import static java.util.Objects.requireNonNull;
+
+import org.jetbrains.java.decompiler.main.DecompilerContext;
+import org.jetbrains.java.decompiler.main.extern.IBytecodeProvider;
+import org.jetbrains.java.decompiler.main.extern.IContextSource;
+import org.jetbrains.java.decompiler.main.extern.IFernflowerLogger;
+import org.jetbrains.java.decompiler.main.extern.IResultSaver;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+final class JarContextSource implements IContextSource, AutoCloseable {
+  private static final String MANIFEST = "META-INF/MANIFEST.MF";
+
+  @SuppressWarnings("deprecation")
+  private final IBytecodeProvider legacyProvider;
+  private final String relativePath; // used for nested contexts from DirectoryContextSource
+  private final File jarFile;
+  private final ZipFile file;
+  private boolean isJar;
+
+  @SuppressWarnings("deprecation")
+  JarContextSource(final IBytecodeProvider legacyProvider, final File archive) throws IOException {
+    this(legacyProvider, archive, "");
+  }
+
+  @SuppressWarnings("deprecation")
+  JarContextSource(final IBytecodeProvider legacyProvider, final File archive, final String relativePath) throws IOException {
+    this.legacyProvider = legacyProvider;
+    this.relativePath = relativePath;
+    this.jarFile = requireNonNull(archive, "archive");
+    this.file = new ZipFile(archive);
+    this.isJar = this.jarFile.getName().endsWith("jar");
+  }
+
+  @Override
+  public String getName() {
+    return "archive " + this.jarFile.getAbsolutePath();
+  }
+
+  @Override
+  public Entries getEntries() {
+    final List<Entry> classes = new ArrayList<>();
+    final List<String> directories = new ArrayList<>();
+    final List<Entry> others = new ArrayList<>();
+
+    Enumeration<? extends ZipEntry> entries = this.file.entries();
+    while (entries.hasMoreElements()) {
+      ZipEntry entry = entries.nextElement();
+
+      String name = entry.getName();
+      if (!entry.isDirectory()) {
+        if (name.endsWith(CLASS_SUFFIX)) {
+          classes.add(Entry.parse(name.substring(0, name.length() - CLASS_SUFFIX.length())));
+        }
+        else if (!this.isJar || !name.equalsIgnoreCase(MANIFEST)) {
+          others.add(Entry.parse(name));
+        }
+      }
+      else {
+        directories.add(name);
```

Comment:
It seems like this doesn't work for all JARs since directory entries are not required to be present

Reply:
Should be fixed as of 9996e7c